include/rdma/fi_domain.h

```diff
@@ -102,6 +104,23 @@ struct fi_ops_av {
 			char *buf, size_t *len);
 	int	(*av_set)(struct fid_av *av, struct fi_av_set_attr *attr,
 			struct fid_av_set **av_set, void *context);
+	int	(*insert_auth_key)(struct fid_av *av, const void *auth_key,
+				   uint64_t flags, void *context);
```

Comment:
Btw, I'm thinking whether we want to return an int to identify an auth_key, a new datatype (fi_auth_t), or reuse fi_addr_t.

The benefits of using fi_addr_t are:

- The caller automatically gets back an any-source address, which is restricted to this auth key.
- We can have a single lookup_auth_key() function (which would use your definition, not my proposed updated one)
- We could re-use the existing AV insert/insertsvc/insertsym calls by having fi_addr_t be an input/output parameter.
- We could also re-use the existing AV lookup/remove calls for handling auth_key's as well as addresses.

Basically, the API update could be limited to adding this call:
```
(*insert_auth_key)(struct fid_av *av, const void *auth_key, size_t auth_key_size,
    fi_addr_t *auth_addr, uint64_t flags);
```
And, honestly, not even that much is needed.  We can use the current insert call with a flag to indicate an auth_key is being inserted.  So, there's some appeal here, but also concern that it might be trying to be too clever.

Note that we can still have new static inline wrappers to make using the API easier.

Reply:
My second comment above was the last one I wrote on the review.  The comments below reference my top comment.  It may help to read the second comment last, since it's presenting a possible alternative that popped into my head after writing the others.

Reply:
>The caller automatically gets back an any-source address, which is restricted to this auth key.

This would remove the need to have an FI_AUTH_KEY_MATCH_ALL flag passed into fi_av_insert() to return a wildcarded EP addr fi_addr_t.

Let me update this proposal taking the fi_addr_t approach instead of the key index approach.